src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,254 @@
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* response) {
+  if (request.service_name() != service_name_) {
+    return Status(Code::INVALID_ARGUMENT,
+                  (string("Invalid service name: ") + request.service_name() +
+                   string(" Expecting: ") + service_name_));
+  }
+
+  if (!request.has_allocate_operation()) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "allocate operation field is required.");
+  }
+
+  if (!cache_) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+  if (!lookup.Found()) {
+    // By returning NOT_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  lookup.value()->Aggregate(request, metric_kinds_.get());
+
+  *response = lookup.value()->check_response();
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// Caches a response from a remote Service Controller AllocateQuota call.
+::google::protobuf::util::Status QuotaAggregatorImpl::CacheResponse(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response) {
+  if (!cache_) {
+    return ::google::protobuf::util::Status::OK;
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  if (cache_) {
+    string request_signature = GenerateAllocateQuotaRequestSignature(request);
+    QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+    CacheElem* cache_elem = new CacheElem(response, SimpleCycleTimer::Now());
+    cache_elem->set_signature(request_signature);
+
+    if (lookup.Found()) {
+      if (lookup.value()->is_refreshing() && lookup.value()->is_aggregated()) {
+        cache_elem->Aggregate(request, metric_kinds_.get());
+      }
+
+      // mark the element is refreshing to avoid
+      // refreshing again bye the cache deleter
+      lookup.value()->set_is_refreshing(true);
+      cache_->Remove(request_signature);
+    }
+
+    // insert aggregated the response to the cache
+    cache_->Insert(request_signature, cache_elem, 1);
+  }
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// When the next Flush() should be called.
+// Returns in ms from now, or -1 for never
+int QuotaAggregatorImpl::GetNextFlushInterval() {
+  if (!cache_) return -1;
+  return options_.expiration_ms;
```

Comment:
Removed expiration_ms. Updated int ServiceControlClientImpl::GetNextFlushInterval() to use this method